cmd/snap/inhibit.go

```diff
@@ -28,22 +28,58 @@ import (
 	"github.com/godbus/dbus"
 	"github.com/snapcore/snapd/cmd/snaplock/runinhibit"
 	"github.com/snapcore/snapd/dbusutil"
+	"github.com/snapcore/snapd/features"
 	"github.com/snapcore/snapd/i18n"
 	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/osutil"
+	"github.com/snapcore/snapd/snap"
 	"github.com/snapcore/snapd/usersession/client"
 )
 
 var runinhibitWaitWhileInhibited = runinhibit.WaitWhileInhibited
 
-func waitWhileInhibited(ctx context.Context, snapName string) error {
+// maybeWaitWhileInhibited is a wrapper for waitWhileInhibited that skips waiting
+// if refresh-app-awareness flag is disabled.
+func maybeWaitWhileInhibited(ctx context.Context, snapName string, appName string) (info *snap.Info, app *snap.AppInfo, hintFlock *osutil.FileLock, err error) {
+	// wait only if refresh-app-awareness flag is enabled
+	if features.RefreshAppAwareness.IsEnabled() {
+		return waitWhileInhibited(ctx, snapName, appName)
+	}
+
+	info, app, err = getInfoAndApp(snapName, appName, snap.R(0))
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	return info, app, nil, nil
+}
+
+// waitWhileInhibited blocks until snap is not inhibited for refresh anymore and then
+// returns a locked hint file lock along with the latest snap and app information.
+// If the snap is inhibited for refresh, a notification flow is initiated during
+// the inhibition period.
+//
+// NOTE: A snap without a hint file is considered not inhibited and a nil FileLock is returned.
+//
+// NOTE: It is the caller's responsibility to release the returned file lock.
+func waitWhileInhibited(ctx context.Context, snapName string, appName string) (info *snap.Info, app *snap.AppInfo, hintFlock *osutil.FileLock, err error) {
 	flow := newInhibitionFlow(snapName)
 	notified := false
+	notInhibited := func(ctx context.Context) (err error) {
+		// Get updated "current" snap info.
+		info, app, err = getInfoAndApp(snapName, appName, snap.R(0))
+		return err
```

Comment:
there is a very small chance that the hint lock is taken and this fails because actually we started without a hint lock file but a refresh started exactly in-between? so if we get a missing current link error in theory we should also retry once

Reply:
I was thinking the same but I now thing this is not possible because soft inhibit check does create the file we flock-locked on.

Reply:
@zyga Actually the soft check doesn't create the hint lock file, only in the hard check (and `backend.UnlinkSnap()`).

Reply:
This definitely a race condition, I added a test for it and fixed in https://github.com/snapcore/snapd/pull/13747/commits/68c29f53b3ce18478866579c176c5be9428dec02